src/mlpack/core/optimizers/smorms3/smorms3_impl.hpp

```diff
@@ -0,0 +1,39 @@
+/**
+ * @file smorms3_impl.hpp
+ * @author Vivek Pal
+ *
+ * Implementation of the SMORMS3 constructor.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_SMORMS3_SMORMS3_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_SMORMS3_SMORMS3_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "smorms3.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template<typename DecomposableFunctionType>
+SMORMS3<DecomposableFunctionType>::SMORMS3(DecomposableFunctionType& function,
+                                           const double stepSize,
+                                           const double epsilon,
+                                           const size_t maxIterations,
+                                           const double tolerance,
+                                           const bool shuffle) :
+    function(function),
```

Comment:
Can you elaborate a bit more? I've tried all I know but this is turning out to be a confusing aspect of meta-programming for me. I have the initialisation as follows:
```cpp
template<typename DecomposableFunctionType>
SMORMS3<DecomposableFunctionType>::SMORMS3(DecomposableFunctionType& function,
                                           const double stepSize,
                                           const double epsilon,
                                           const size_t maxIterations,
                                           const double tolerance,
                                           const bool shuffle) :
    optimizer(SGD<DecomposableFunctionType, SMORMS3Update>(function,
                                                           stepSize,
                                                           maxIterations,
                                                           tolerance,
                                                           shuffle,
                                                           SMORMS3Update))
{ /* Nothing to do. */ }
```
but it results in a couple of [errors](https://paste.debian.net/922572/) and a warning for unused `epsilon`. Can you please take a look at those errors?